src/main/kotlin/org/rust/ide/hints/RsGenericParameterInfoHandler.kt

```diff
@@ -0,0 +1,148 @@
+import com.intellij.lang.parameterInfo.*
+import com.intellij.openapi.util.TextRange
+import com.intellij.util.containers.nullize
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+
+private const val WHERE_PREFIX = "where "
+
+class RsGenericParameterInfoHandler : ParameterInfoHandler<RsTypeArgumentList, RsGenericPresentation> {
+
+    var curParam = -1
+        private set
+    var hintText = ""
+        private set
+
+    override fun showParameterInfo(element: RsTypeArgumentList, context: CreateParameterInfoContext) {
+        context.highlightedElement = null
+        context.showHint(element, element.textRange.startOffset, this)
+    }
+    
+    // todo: don't disappear in nested generic types
+    override fun updateParameterInfo(parameterOwner: RsTypeArgumentList, context: UpdateParameterInfoContext) {
+        curParam = ParameterInfoUtils.getCurrentParameterIndex(parameterOwner.node, context.offset, RsElementTypes.COMMA)
+        when {
+            context.parameterOwner == null -> context.parameterOwner = parameterOwner
+
+            // occurs e.g. in case of up-down cursor moving
+            context.parameterOwner != parameterOwner -> context.removeHint()
+        }
+    }
+
+    override fun updateUI(p: RsGenericPresentation, context: ParameterInfoUIContext) {
+        hintText = p.presentText
+        val (startOffset, endOffset) =
+            if (hintText.startsWith(WHERE_PREFIX))
+                0 to WHERE_PREFIX.length
+            else
+                p.getRange(curParam).startOffset to p.getRange(curParam).endOffset
+        context.setupUIComponentPresentation(
+            hintText,
+            startOffset,
+            endOffset,
+            false, // grayed
+            false,
+            false, // define grayed part of args before highlight
+            context.defaultParameterColor
+        )
+    }
+
+    override fun getParametersForLookup(item: LookupElement?, context: ParameterInfoContext?): Array<Any>? = null
+
+    override fun couldShowInLookup() = false
+
+    override fun findElementForUpdatingParameterInfo(context: UpdateParameterInfoContext): RsTypeArgumentList? =
+        findExceptColonColon(context)
+
+    override fun findElementForParameterInfo(context: CreateParameterInfoContext): RsTypeArgumentList? {
+        val parameterList = findExceptColonColon(context) ?: return null
+        val parent = parameterList.parent
+        val genericDeclaration = when (parent) {
+            is RsMethodCall,
+            is RsPath -> parent.reference?.resolve()
+            else -> return null
+        } as? RsGenericDeclaration ?: return null
+        val typesWithBounds = genericDeclaration.typeParameters.nullize() ?: return null
+        context.itemsToShow = listOfNotNull(
+            RsGenericPresentation(typesWithBounds, false),
+            RsGenericPresentation(typesWithBounds, true))
+            .filterNot { it.presentText == "" }
```

Comment:
Use `isEmpty` instead if possible

Reply:
These hack with true/false params and filters looks like a very bad design.